torch/nn/utils/stateless.py

```diff
@@ -37,144 +45,169 @@ def _setattr(self, name: str, value: Any) -> None:
+            # Detect if there are multiple keys present for the same tied tensor.
+            len(tied_names.intersection(given_names_for_tied_tensors)) > 1
+            # Only raise an error if the user passed multiple values for the same tied tensor.
+            # If all given values are the same, don't raise.
+            and len({parameters_and_buffers[tied_name] for tied_name in tied_names})
+            != 1
+        ):
+            raise ValueError(
+                f"functional_call got multiple values for keys {sorted(tied_names)}, "
+                f"which are tied. Consider using tie_weights=False"
+            )
 
-def _remove_swap(module, name: str, full_path: str) -> None:
-    if hasattr(module, "_orig_class"):
-        module.__class__ = module._orig_class
-        delattr(module, "_orig_class")
-        delattr(module, "_attr_to_path")
+    # Untie the given named tensor map
+    # Make a copy for not modifying the original dict
+    untied_parameters_and_buffers = parameters_and_buffers.copy()
+    for given_name in given_names_for_tied_tensors:
+        for tied_name in tied_names_map[given_name]:
+            untied_parameters_and_buffers[tied_name] = parameters_and_buffers[
+                given_name
+            ]
+    return untied_parameters_and_buffers
 
 
 @contextlib.contextmanager
 def _reparametrize_module(
-    module: 'torch.nn.Module',
+    module: "torch.nn.Module",
     parameters_and_buffers: Dict[str, Tensor],
+    *,
     tie_weights: bool = False,
+    strict: bool = False,
 ) -> Iterator[None]:
-    tied_weights_map = _create_tied_weights_map(module, parameters_and_buffers) if tie_weights else {}
-    for name, tensor in parameters_and_buffers.items():
-        _apply_func_submodules(
-            _create_swap_params(parameters_and_buffers),
-            module, name.split("."), name, (tensor,))
-    for tied_name, user_given_name in tied_weights_map.items():
-        _apply_func_submodules(
-            _create_swap_params(parameters_and_buffers),
-            module, tied_name.split("."), user_given_name, (None,))
+    if tie_weights:
+        untied_parameters_and_buffers = _untie_named_tensors_map(
+            module, parameters_and_buffers
+        )
+    else:
+        untied_parameters_and_buffers = parameters_and_buffers
+
+    accessor = NamedMemberAccessor(module)
+    if strict:
+        missing_keys, unexpected_keys = accessor.check_keys(
+            untied_parameters_and_buffers
+        )
+        error_msgs = []
+        if len(unexpected_keys) > 0:
+            error_msgs.append(
+                "Unexpected key(s): {}.".format(", ".join(map(repr, unexpected_keys)))
+            )
+        if len(missing_keys) > 0:
+            error_msgs.append(
+                "Missing key(s): {}.".format(", ".join(map(repr, missing_keys)))
+            )
+        if len(error_msgs) > 0:
+            raise RuntimeError(
+                "Error(s) in reparameterizing for {}:\n\t{}".format(
```

Comment:
```suggestion
                "Error(s) in reparametrizing for {}:\n\t{}".format(
```

Reply:
I'm fine with both `reparameterize` and `reparametrize`. Both spellings are correct and both are used in PyTorch. E.g., `torch.distributions` uses "parameterize" (/w "e") while in `torch.nn` and `torch.aot_autograde`, we use "parametrize" (/wo "e").

"parameterize" (/w "e") looks more common.

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/16078332/216742681-ab8976ab-8091-4e75-89fc-9353703c5dbb.png">

Reply:
The general "rule" we use for this (to try and preserve consistency):
- parameterize means turning something into a parameter.
- parametrize means reformulating something into something else via parameters (parametrize a curve).

Not that these are necessarily correct but at least that allows us to (try) to be consistent within PyTorch).